packages/babel-parser/src/parser/expression.js

```diff
@@ -1158,7 +1170,7 @@ export default class ExpressionParser extends LValParser {
     this.next();
     meta = this.createIdentifier(meta, "function");
 
-    if (this.scope.inGenerator && this.eat(tt.dot)) {
+    if (this.prodParam.hasYield && this.eat(tt.dot)) {
```

Comment:
2x member access is expensive.

Reply:
Yeah, but `this.prodParam.hasYield` is still faster than `scope.inGenerator` because we don't have to walkup the scope stacks to find a var scope:

https://github.com/babel/babel/blob/09cb4273bc74e1916db17920f068cad563eefd00/packages/babel-parser/src/util/scope.js#L209

Reply:
What about a mutual parser flag instead?  `state.flag |= HasYield` and then 
`if (state.flag & HasYield) ...``

Reply:
Do you think that it would make any difference? I have always assumed that we have way bigger perf problems in our parser, and that those optimizations would only be noticeable after we resolve things like "Every method access has two walk different levels of the prototype chain" and "Almost every function is megamorphic" and "Our AST objects continuously change their shape".

Reply:
`this.prodParam.hasYield` is implemented by parser flags. V8 will inline these Get accessors eventually.

The scope stack is used to track nested production params. It is equivalent to do things like
```js
const oldHasYield = this.state.hasYield
this.state.hasYield = true;
parseFunctionBody()
this.state.hasYield = oldHasYield;
```
The only difference is that the `hasYield` above is stored implicitly in stack while `this.prodParams.hasYield` stored explicitly in heap.

Reply:
@nicolo-ribaudo If you see the whole picture, maybe no :) There are too many bottle necks in this parser such as the lexer, tokens itself etc. 

Why not develop a new parser from bottom up? It shouldn't take more than 10 days (estimated) to replicate current Babel parser with working plugins.

Reply:
If you see the even bigger picture, then it would probably be more valuable to re-implement `@babel/traverse` which is where Babel spends most of its time :joy:

Reply:
Just looked at `@babel/traverse`. Should be easy :) Just get rid of the prototype and use functions instead, plus use bitmasks to handle the flow.
The `findParent` function is one of the root causes I guess for why it's slow :)